src/Datadog.Trace/Configuration/TracerSettings.cs

```diff
@@ -244,7 +247,23 @@ public TracerSettings(IConfigurationSource source)
         /// <summary>
         /// Gets or sets the map of header keys to tag names, which are applied to the root <see cref="Span"/> of incoming requests.
         /// </summary>
-        public IDictionary<string, string> HeaderTags { get; set; }
+        public IDictionary<string, string> HeaderTags
+        {
+            get => _headerTags;
+            set
+            {
+                _headerTags = value;
+
+                if (value is ConcurrentDictionary<string, string> concurrentDictionary)
+                {
+                    _headerTagsEmpty = () => concurrentDictionary.IsEmpty;
```

Comment:
Trying to understand the delegate + closure here, It's the delegate optimisation done to avoid calling `.Count` of the concurrent dictionary?

Reply:
Yup. `Count` always acquires the lock, whereas `IsEmpty` is lock-free as long as the dictionary is empty (which is the case I'm optimizing for).

The closure could be avoided, but since it's only allocated when initializing the settings (so once or twice in total), I figured it wouldn't make much of a difference.

Reply:
Ok, and can't you just have that logic in the `HeaderTagsEmpty` property? like:
```csharp
return HeaderTags is ConcurrentDictionary<string, string> concurrentDictionary ? concurrentDictionary.IsEmpty : HeaderTags.Count == 0;
```

Or is the casting a big performance reason to remove it?

Reply:
You're right, performance gain is negligible, gonna go for the simpler cast